src/rgw/rgw_op.cc

```diff
@@ -1918,6 +2062,16 @@ int RGWListBucket::verify_permission()
     return -EACCES;
   }
 
+  // TODO: Is this correct?
+  if (s->iam_policy) {
+    if (s->iam_policy->eval(s->env, *s->auth.identity,
+			    s->object.instance.empty() ?
+			    rgw::IAM::s3ListBucket :
+			    rgw::IAM::s3ListBucketVersions,
```

Comment:
this check should be based on the `RGWListBucket::list_versions` flag. the only problem is that it's assigned later in the request processing (when `RGWListBucket_ObjStore_S3::get_params()` is called by `RGWListBucket::execute()`), so you won't have access to it here. you probably want to just check `s->info.args.exists("versions")`